L74-java-channel-creds.md

```diff
@@ -0,0 +1,478 @@
+}
+```
+
+Unlike most of the credentials, TLS may have a considerable amount of
+configuration. And as time goes on, more configuration options may be provided.
+It could be a severe bug if certain TLS restrictions were ignored by a channel.
+To detect cases where the channel does not understand enough of the TLS
+configuration, there will be a method to check if there are required features
+that are not understood. If some are not understood, enough information will be
+returned to produce a useful error message.
+
+`TlsChannelCredentials` will support a no-arg constructor for the "all
+defaults" configuration. For any more advanced configuration, a builder would
+be used.
+
+```java
+package io.grpc;
+
+public final class TlsChannelCredentials extends ChannelCredentials {
+  /**
+   * Returns an empty set if this credential can be adequately understood via
+   * the features listed, otherwise returns a hint of features that are lacking
+   * to understand the configuration to be used for manual debugging.
+   *
+   * <p>An "understood" feature does not imply the caller is able to fully
+   * handle the feature. It simply means the caller understands the feature
+   * enough to use the appropriate APIs to read the configuration. The caller
+   * may support just a subset of a feature, in which case the caller would
+   * need to look at the configuration to determine if only the supported
+   * subset is used.
+   *
+   * <p>This method may not be as simple as a set difference. There may be
+   * multiple features that can independently satisfy a piece of configuration.
+   * If the configuration is incomprehensible, all such features would be
+   * returned, even though only one may be necessary.
+   *
+   * <p>An empty set does not imply that the credentials are fully understood.
+   * There may be optional configuration that can be ignored if not understood.
+   */
+  public Set<Feature> incomprehensible(EnumSet<Feature> understoodFeatures) {...}
+
+  /* Other methods can be added to support various Features */
+
+  public enum Feature {}
+
+  public final static class Builder {
+    public TlsChannelCredentials build() {...}
+  }
+}
+```
+
+To support very advanced use cases, Netty will provide a credential that wraps
+a `ProtocolNegotiator`. This allows implementations like ALTS and XDS to use
+internal APIs without forcing their users to use experimental or internal APIs,
+as their users would just interact with `ChannelCredentials`.
+
+```java
+package io.grpc.netty;
+
+final class NettyChannelCredentials extends ChannelCredentials {
+  public NettyChannelCredentials(ProtocolNegotiator.Factory negotiator) {...}
+  public ProtocolNegotiator.Factory getNegotiator() { return negotiator; }
+}
+
+@Internal
+public final class InternalNettyChannelCredentials {
+  private InternalNettyChannelCredentials() {}
+
+  public ChannelCredentials create(InternalProtocolNegotiator.Factory negotiator)
```

Comment:
not sure where you draw the boundary for credentials, but this would be a pretty big expansion.     This API is powerful to pick things like ciphersuites, TLSv1.3, session tickets, SNI support (server side), etc.  Those are all security related, but don't feel like Credentials. (e.g. keystores and truststores)

Reply:
I was expecting users to continue providing Netty's SslContext. I will need some way to convert an SslContext to ChannelCredentials (whose concrete type is InternalNettyChannelCredentials). But yeah, there is a missing API here. Make a TlsNettyChannelCredentials container class for a static utility method? I'll have to think about it a bit.

Reply:
IIUC, even if we are expecting users to provide SslContext, we still don't lose the ability to extend the API to be able to pick some security-related things, right? For example, in C core I know there are some ongoing work to choose between TLSv1.3 and TLSv1.2 through credential options(maybe @jiangtaoli2016 can provide any additional information), if one day we decide to do that in Java too, I guess we can easily do that as well?

Reply:
We can create new netty negotiators at any time and expose them with an API. They would have full reign over what they want to do.

I'm not sure what we'd want to do about TLSv1.2 vs TLSv1.3; on that front I'd expect we'd just start supporting TLSv1.3 as soon as it works. Deciding when to disable TLSv1.2 may need a Feature in TlsChannelCredentials (or maybe it is optional, only observed in the transport supports it).

Reply:
I've added notes about `NettySslContextChannelCredentials` API to convert a netty `SslContext` to `ChannelCredentials`.